expfmt/openmetrics_create.go

```diff
@@ -0,0 +1,503 @@
+// Copyright 2014 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package expfmt
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"math"
+	"strconv"
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/prometheus/common/model"
+
+	dto "github.com/prometheus/client_model/go"
+)
+
+// MetricFamilyToOpenMetrics converts a MetricFamily proto message into the
+// OpenMetrics text format and writes the resulting lines to 'out'. It returns
+// the number of bytes written and any error encountered. The output will have
+// the same order as the input, no further sorting is performed. Furthermore,
+// this function assumes the input is already sanitized and does not perform any
+// sanity checks. If the input contains duplicate metrics or invalid metric or
+// label names, the conversion will result in invalid text format output.
+//
+// This method fulfills the type 'prometheus.encoder'.
+//
+// Note that OpenMetrics requires a final `# EOF` line. Since this function acts
+// on individual metric families, it is the responsibility of the caller to
+// append this line to 'out' once all metric families have been written.
+//
+// TODO(beorn7): The output is not yet fully OpenMetrics compliant. It should,
+// however, work with Prometheus. Known deficiencies are:
+//
+// - A `_total` suffix for counters is not enforced, but if a counter has it, it
+//   is truncated from the `# TYPE` and `# HELP` line.
+//
+// - No support for `# UNIT` lines, info type, enum type, `_created` lines.
```

Comment:
None of these are mandatory. s/enum/stateset/, and gaugehistogram is missing too.

Reply:
I have changed the wording. With invalidly named counters changed to `unknown`, the output should now be strictly OM compliant. I changed the wording in this regard, too.